File: yattag.py
#!/usr/bin/env python3
#
# Copyright (c) 2021 Miklos Vajna and contributors.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""
The yattag module generates HTML with Python.

This is a stripped down version of the Python package.
"""

from typing import Any
from typing import List
from typing import Tuple


class Doc:
    """Generates xml/html documents."""
    def __init__(self) -> None:
        self.value = ""

    def get_value(self) -> str:
        """Gets the escaped value."""
        return self.value

    def append_value(self, value: str) -> None:
        """Appends escaped content to the value."""
        self.value += value

    def tag(self, name: str, attrs: List[Tuple[str, str]]) -> 'Tag':
        """Starts a new tag."""
        return Tag(self, name, attrs)

    def stag(self, name: str, attrs: List[Tuple[str, str]]) -> None:
        """Starts a new tag and closes it as well."""
        self.append_value("<{}".format(name))
        for attr in attrs:
            key = attr[0]
            value = attr[1].replace("&", "&amp;").replace("<", "&lt;").replace('"', "&quot;")
            self.append_value(" {}=\"{}\"".format(key, value))
        self.append_value(" />")

    def text(self, text: str) -> None:
        """Appends unescaped content to the document."""
        self.append_value(text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;"))


class Tag:
    """Starts a tag, which is closed automatically."""
    def __init__(self, doc: Doc, name: str, attrs: List[Tuple[str, str]]) -> None:
        doc.append_value("<{}".format(name))
        for attr in attrs:
            key = attr[0]
            value = attr[1].replace("&", "&amp;").replace("<", "&lt;").replace('"', "&quot;")
            doc.append_value(" {}=\"{}\"".format(key, value))
        doc.append_value(">")
        self.doc = doc
        self.name = name

    def __enter__(self) -> None:
        pass

    def __exit__(self, tpe: Any, value: Any, traceback: Any) -> None:
        self.doc.append_value("</{}>".format(self.name))


# vim:set shiftwidth=4 softtabstop=4 expandtab:

File: src/yattag.rs
/*
 * Copyright 2021 Miklos Vajna. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#![deny(warnings)]
#![warn(clippy::all)]
#![warn(missing_docs)]

//! Generate HTML with Rust.
//!
//! This is more or less a Rust port of <https://www.yattag.org/>, mostly because
//! <https://crates.io/crates/html-builder> would require you to manually escape attribute values.

use std::cell::RefCell;
use std::fmt::Write as _;
use std::rc::Rc;

/// Generates xml/html documents.
#[derive(Clone)]
pub struct Doc {
    value: Rc<RefCell<String>>,
}

impl Doc {
    /// Creates an empty Doc.
    pub fn new() -> Doc {
        Doc {
            value: Rc::new(RefCell::new(String::from(""))),
        }
    }

    /// Factory of yattag.Doc from a string.
    pub fn from_text(text: &str) -> Self {
        let doc = Doc::new();
        doc.text(text);
        doc
    }

    /// Gets the escaped value.
    pub fn get_value(&self) -> String {
        self.value.borrow().clone()
    }

    /// Appends escaped content to the value.
    pub fn append_value(&self, value: String) {
        self.value.borrow_mut().push_str(&value)
    }

    /// Starts a new tag.
    pub fn tag(&self, name: &str, attrs: &[(&str, &str)]) -> Tag {
        Tag::new(&self.value, name, attrs)
    }

    /// Starts a new tag and closes it as well.
    pub fn stag(&self, name: &str) {
        self.append_value(format!("<{} />", name));
    }

    /// Appends unescaped content to the document.
    pub fn text(&self, text: &str) {
        let encoded = html_escape::encode_safe(text).to_string();
        self.append_value(encoded);
    }
}

/// HtmlTable is a matrix (rows, then cols) of Doc instances.
pub type HtmlTable = Vec<Vec<Doc>>;

impl Default for Doc {
    fn default() -> Self {
        Self::new()
    }
}

/// Starts a tag, which is closed automatically.
pub struct Tag {
    value: Rc<RefCell<String>>,
    name: String,
}

impl Tag {
    fn new(value: &Rc<RefCell<String>>, name: &str, attrs: &[(&str, &str)]) -> Tag {
        let mut guard = value.borrow_mut();
        guard.push_str(&format!("<{}", name));
        for attr in attrs {
            let key = attr.0;
            let val = html_escape::encode_double_quoted_attribute(&attr.1);
            guard.push_str(&format!(" {}=\"{}\"", key, val));
        }
        guard.push('>');
        let value = value.clone();
        Tag {
            value,
            name: name.to_string(),
        }
    }

    /// Appends unescaped content inside a tag.
    pub fn text(&self, text: &str) {
        let encoded = html_escape::encode_safe(text).to_string();
        self.value.borrow_mut().push_str(&encoded)
    }

    /// Starts a new tag inside a tag.
    pub fn tag(&self, name: &str, attrs: &[(&str, &str)]) -> Tag {
        Tag::new(&self.value, name, attrs)
    }

    /// Starts a new tag and closes it as well, inside a tag.
    pub fn stag(&self, name: &str, attrs: &[(&str, &str)]) {
        self.append_value(format!("<{}", name));
        for attr in attrs {
            let key = attr.0;
            let value = html_escape::encode_double_quoted_attribute(&attr.1);
            self.append_value(format!(" {}=\"{}\"", key, value));
        }
        self.append_value(String::from("/>"))
    }

    /// Appends escaped content inside a tag.
    pub fn append_value(&self, value: String) {
        self.value.borrow_mut().push_str(&value)
    }
}

impl Drop for Tag {
    fn drop(&mut self) {
        let _ = write!(self.value.borrow_mut(), "</{}>", self.name);
    }
}

#[cfg(test)]
mod tests;
